File: leetcode/solutions/0345.ReverseVowelsOfAString/reverseVowels.go
package solution

func reverseVowels(s string) string {
	str := []rune(s)
	left, right := 0, len(str)-1

	for ; left < right; left, right = left+1, right-1 {
		for ; left < right && !vowels[str[left]]; left++ {
		}
		for ; left < right && !vowels[str[right]]; right-- {
		}

		str[left], str[right] = str[right], str[left]
	}

	return string(str)
}

var vowels = map[rune]bool{
	'A': true,
	'a': true,
	'I': true,
	'i': true,
	'U': true,
	'u': true,
	'E': true,
	'e': true,
	'O': true,
	'o': true,
}

File: leetcode/solutions/0345.ReverseVowelsOfAString/reverseVowels.ts
export default function reverseVowels(s: string): string {
  const str = s.split("");
  let left = 0;
  let right = str.length - 1;

  for (; left < right; left++, right--) {
    for (; left < right && !VOWELS.has(str[left]); left++) {}
    for (; left < right && !VOWELS.has(str[right]); right--) {}

    const tmp = str[left];
    str[left] = str[right];
    str[right] = tmp;
  }

  return str.join("");
}

const VOWELS = new Set<string>([
  "A",
  "a",
  "I",
  "i",
  "U",
  "u",
  "E",
  "e",
  "O",
  "o",
]);
